wdl-lint/src/rules/shellcheck.rs

```diff
@@ -0,0 +1,387 @@
+        section: &CommandSection,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        if !SHELLCHECK_EXISTS.get_or_init(|| {
+            if !program_exists(SHELLCHECK_BIN) {
+                let command_keyword = support::token(section.syntax(), SyntaxKind::CommandKeyword)
+                    .expect("should have a command keyword token");
+                state.exceptable_add(
+                    Diagnostic::error("running `shellcheck` on command section")
+                        .with_label(
+                            "could not find `shellcheck` executable.",
+                            command_keyword.text_range().to_span(),
+                        )
+                        .with_rule(ID)
+                        .with_fix("install shellcheck or disable this lint."),
+                    SyntaxElement::from(section.syntax().clone()),
+                    &self.exceptable_nodes(),
+                );
+                return false;
+            }
+            true
+        }) {
+            return;
+        }
+
+        // Collect declarations so we can ignore placeholder variables
+        let parent_task = section.parent().into_task().expect("parent is a task");
+        let mut decls = gather_task_declarations(&parent_task);
+
+        // Replace all placeholders in the command with dummy bash variables
+        let Some((sanitized_command, cmd_decls)) = sanitize_command(section) else {
+            // This is the case where the command section contains
+            // mixed indentation. We silently return and allow
+            // the mixed indentation lint to report this.
+            return;
+        };
+        decls.extend(cmd_decls);
+
+        // Get leading whitespace so we can add it to each span
+        let leading_whitespace = count_command_whitespace(section);
+
+        // Map each actual line of the command to its corresponding
+        // `CommandPart` and start position.
+        let mut line_map = HashMap::new();
+        let mut line_num = 1;
+        let mut on_same_line = false;
+        for part in section.parts() {
+            match part {
+                CommandPart::Text(ref text) => {
+                    for (line, start, _) in lines_with_offset(text.as_str()) {
+                        if line_num == 1 && line.trim().is_empty() {
+                            continue;
+                        }
+                        if on_same_line {
+                            on_same_line = false;
+                            continue;
+                        }
+                        line_map.insert(
+                            line_num,
+                            text.span().start() + start + leading_whitespace - 1,
+                        );
+                        line_num += 1;
+                    }
+                }
+                CommandPart::Placeholder(_) => {
+                    on_same_line = true;
+                }
+            }
+        }
```

Comment:
I don't think this algorithm will work if there's wonky leading whitespace?
e.g.
```
command <<<     # this is funky
         # more weirdness
echo ~{greeting}
     echo ~{name}
   >>>
```

Regardless, I think we need some more documentation for this. I would move this block to its own function, and write a doc comment explaining what it does in detail. Just skimming this, I'm not as clear on what's happening as I'd like.

Reply:
I think multi-line placeholders will also screw it up, if I'm reading correctly.
e.g.
```
command <<<
    python script.py ~{(
        if bool
        then complex
        else expression
    )} -more -flags -all-one-same-line
>>>
```

Reply:
Concerns should be addressed with the corrected algorithm now committed. Tests now include a variety of weirdness and multiline placeholders.